scripts/o3de/o3de/utils.py

```
@@ -348,6 +456,22 @@ def find_ancestor_dir_containing_file(target_file_name: pathlib.PurePath, start_
     ancestor_file = find_ancestor_file(target_file_name, start_path, max_scan_up_range)
     return ancestor_file.parent if ancestor_file else None
 
+def get_project_path_from_file(target_file_path: pathlib.Path, supplied_project_path: pathlib.Path = None) -> pathlib.Path:
+    """
+    Based on a file supplied by the user, and optionally a differing project path, determine and validate a proper project path, if any.
+    :param target_file_path: A user supplied file path
+    :param supplied_project_path: (Optional) If the target file is different from the project path, the user may also specify this path as well.
+    :return: A valid project path, or None
+    """
+    project_path = supplied_project_path
+    if not project_path:
+        project_path = find_ancestor_dir_containing_file(pathlib.PurePath('project.json'), target_file_path)
+        if not project_path:
+            raise valid.InvalidProjectPathException(f"Unable to find project folder associated with file '{target_file_path}'. Please specify using --project-path, or ensure the file is inside a project folder.")
```

There is nothing about this function I believe should throw an exception.

As this is a utility function, it can just `return None` if the target path isn't under the project.

By throwing the exception, the function is expecting in the normal call patterns that the target_file_path is always underneath a project_path, which doesn't seem like a reasonable expectation for a utilty.

This intent of this function seems more to return the project_root if it a file is underneath the project, other return none.

---

I would like to retain relevant error messages, because it can help troubleshoot what is missing.
Would you be okay with a convention where the function returns a tuple of the result and a string encapsulating the error message (if any)?

If not, what would you recommend for relaying error messages? Should that be handled by this function, or the caller?

---

The tuple convention works, and would be inline with how we use AZ::Outcome(std::expected) in C++.

My point here is that some of depending on the operation the function is meant to perform some of the actions aren't errors.
For example if the caller invokes a a function to open a file at path `<project-path>/foo`, could be an error that can throw an exception if the file doesn't exist.
But if the caller is invoking a function to find the nearest ancestor project path from a source path, then not finding the path isn't an error, it is within the expectation of the function.